src/RazorSdk/SourceGenerators/IncrementalValueProviderExtensions.cs

```diff
@@ -8,8 +8,43 @@
 
 namespace Microsoft.NET.Sdk.Razor.SourceGenerators
 {
+    internal sealed class RazorSourceGeneratorComparer<T> : IEqualityComparer<(T Left, RazorSourceGenerationOptions Right)> where T : notnull
+    {
+        private readonly Func<T, T, bool> _equals;
+        public RazorSourceGeneratorComparer(Func<T, T, bool>? equals = null)
+        {
+            _equals = equals ?? EqualityComparer<T>.Default.Equals;
+        }
+
+        public bool Equals((T Left, RazorSourceGenerationOptions Right) x, (T Left, RazorSourceGenerationOptions Right) y)
+        {
+            if (y.Right.SuppressRazorSourceGenerator)
+            {
+                // If source generation is suppressed, we can always use previously cached results.
+                return true;
```

Comment:
>return true; [](http://example.com/codeflow?start=16&length=12)

Does this mean `RazorSourceGeneratorComparer<T>` is not symmetric? That is `Equals(x, y)` does not necessarily imply `Equals(y, x)`?

Reply:
Yes, its actually designed for that to specifically be the case. Any time the options flip from unsuppressed -> suppressed we want to override and pretend everything is cached and do no work. When we switch back to unsuppressed we want to do the actual check.  It's taking advantage of knowing how the generator internals work (https://sourceroslyn.io/#Microsoft.CodeAnalysis/SourceGeneration/Nodes/NodeStateTable.cs,386), so a a horrible hack frankly but necessary to support razors odd behavior caused by the design time / runtime split (something we're working on removing which would remove all these required hacks). 

> And does Equals(x, y) guarantee GetHashCode(x) == GetHashCode(y)

It does not, but again we know this will never actually be used in a hash table, so it's never called. I think I'll add some comments on this to make it clear its a very specialized piece of code that shouldn't be re-used generally.